test/Streams/SeekableReader.test.cpp

```diff
@@ -0,0 +1,65 @@
+#include "Streams/MemoryReader.h"
+#include "Streams/FileReader.h"
+#include "Streams/FileSliceReader.h"
+#include <gtest/gtest.h>
+#include <array>
+
+template <class T>
+Stream::SeekableReader* CreateSeekableReader();
+
+template <>
+Stream::SeekableReader* CreateSeekableReader<Stream::MemoryReader>() {
+	std::array<char, 5> buffer{ 't', 'e', 's', 't', '!' };
+	return new Stream::MemoryReader(buffer.data(), buffer.size() * sizeof(char));
+}
+
+template <>
+Stream::SeekableReader* CreateSeekableReader<Stream::FileReader>() {
+	return new Stream::FileReader("Streams/SimpleStream.txt");
+}
+
+template <>
+Stream::SeekableReader* CreateSeekableReader<Stream::FileSliceReader>() {
+	Stream::FileReader fileReader("Streams/SimpleStream.txt");
+	return &fileReader.Slice(5);
+}
+
+// Define a test fixture class template
+template <class T>
+class SimpleSeekableReader : public testing::Test {
+protected:
+	// The ctor calls the factory function to create a reader implemented by T
+	SimpleSeekableReader() : seekableReader(CreateSeekableReader<T>()) {}
+
+	virtual ~SimpleSeekableReader() { delete seekableReader; }
+
+	Stream::SeekableReader* const seekableReader;
+	const unsigned int size = 5;
+};
+
+using testing::Types;
+
+typedef Types<Stream::MemoryReader, Stream::FileReader> SeekableStreamImplementations;
```

Comment:
This requires seeing the definitions of all stream types at the same time. That's maybe not the best design, especially since library users could define their own streams, but might still want to make use of this test suite to verify their own implementations.

I think this is where `REGISTER_TYPED_TEST_CASE_P` comes in handy. We can collect all the tests together into a test suit, and let them instantiate it for individual types in their own test files using `INSTANTIATE_TYPED_TEST_CASE_P`. It's slightly more work on our end to manually register all the tests into a test suite with the register macro, but I think the benefits are strongly worth it.

Reply:
I would like to point out this would still be extendable if the user was to fork the library and just add their implementation to the list of types. I don't have any issues with trying the REGISTER_TYPED_TEST_CASE_P implementation although I haven't learned enough to try it yet.

Reply:
Ahh, because such an edit requires forking the library. It shouldn't be necessary to fork a library to extended the classes and interfaces it provides. This is much like how designing a new Iterator doesn't and shouldn't require forking the standard C++ library.

For a more concrete motivating example, suppose someday the Stream section is separated out to it's own library. Now suppose the VOL decompression code was rewritten to conform to the Stream interface. Where would decompression code go for VOL files? Probably not in the Stream library. It would probably remain with the rest of the VOL processing code, which would import the Stream code. So how then are these decompression streams tested? Hopefully not by editing the test code within the Stream library, but by importing the tests from the Stream library into the test project for the VOL processing code.

This is what `REGISTER_TYPED_TEST_CASE_P` allows us to do.

Reply:
All sounds good. I just felt like the GoogleTest documentation was a little draconian in saying that only the second method was suitable for extending outside of the library.

Reply:
I think it's about technical limitations. With the first method, the list of types must be specified before the tests. Hence there is no way to import the tests from the library and then specify the types later. When playing around with the original typed test code, I tried moving the type instantiations to after the tests and it was considered an error. There's like a technical reason for that buried in the macros.

With the type-parameterized tests, I think they have problems wrapping all necessary parts into a single macro. Hence the separate declaration of the test case above, and then manually registering all tests in the test case below. It makes it less pleasant to define tests that way. You lose out on the auto registration of tests. I suppose that's why they offer both methods, so you don't pay the cost of the extra setup when you don't need it.

----

Maybe there is a way with the first method to define the types first, and then import the library. Though that's an awkward backwards way of doing things, which seems non-standard and surprising. Personally, I like the second method from a library user's standpoint. The hard stuff has been taken care of in the library, so it's natural and easy for library users to use and extend.